### PR TITLE
hal/armv7a: cache management in pmap

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

Cache management is added to `pmap_enter` and `pmap_remove` to provide data consistency. According to: https://support.xilinx.com/s/article/52035?language=en_US TTL1 and TTL2 entries have to be cleaned and invalidated before update.

Moreover:
1. Barriers have been added to cp15 coprecossor, according to ARM documentation.
2. Added some comments and defines in pmap.

**TODO:**
In current approach TLB is invalidated after each `pmap_switch`, it is unnecessary. ASIDs mechanisms allows to have the same virtual addresses from different processes in TLB. TLB invalidation should be performed in asid alloc/dealloc function when the asids pool is run out.
 
**CHANGES HAVE NOT BEEN TESTED ON DTR PROJECT**
## Motivation and Context

Executing programs from file system caused abort or prefetch exceptions:
image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (zynq7000, qemu-zynq7000, imx6ull-devkit).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/armv7a/pmap.c [hunk 1]
```diff
@@ -231,77 +233,82 @@ void pmap_switch(pmap_t *pmap)
...
 	hal_cpuInvalASID(pmap_common.asids[pmap->asid_ix]);
+
 	hal_cpuDataSyncBarrier();
 	hal_cpuInstrBarrier();
 }
 
 
 static void _pmap_mapScratch(addr_t pa, unsigned char asid)
 {
+	/* Entry into global kernel page table, hence asid = 0 */
```
Comment (agkaminski) [inline hal/armv7a/pmap.c:276]: You forgot to set asid to zero below

File: hal/armv7a/_armv7a.S [hunk 2]
```diff
@@ -26,111 +26,187 @@ hal_cpuGetCycles:
...
 hal_cpuInvalDataCache:
-	ldr r1, =SIZE_CACHE_LINE - 1
-	bic r0, r0, r1
-	mcr p15, 0, r0, c7, c6, 1
+	dsb
+	mrc p15, 0, r3, c0, c0, 1 /* Read the Cache Type Register */
+	lsr r3, r3, #16           /* DMinLine value */
+	and r3, r3, #0xf
+	mov r2, #4
+	mov r2, r2, lsl r3        /* Cache line size in bytes */
```
[the remark this answers is not in the document]
Comment (kemonats) [reply hal/armv7a/_armv7a.S:66]: I guess not
```
ASM hal/armv7a/zynq7000/_init.S
hal/armv7a/zynq7000/../_armv7a.S: Assembler messages:
hal/armv7a/zynq7000/../_armv7a.S:65: Error: constant expression expected -- `mov r2,#4,lsl r3'
```
ref
[ARM mov](https://developer.arm.com/documentation/dui0489/i/arm-and-thumb-instructions/movl)

**Review by agkaminski — APPROVED** [review]
LGTM

File: hal/armv7a/pmap.c (hunk 1 above)
Comment (HBuczynski) [reply hal/armv7a/pmap.c]: The comment has been removed.

**Pull request merged**